.dev/scripts/generate_changelog.py

```diff
@@ -0,0 +1,233 @@
+                data["message"]
+            )
+        )
+    return data["created_at"]
+
+
+def fetch_pr_data(start_date):
+    """
+    Fetch data of PRs merged after a given start date.
+    """
+    url = "https://api.github.com/search/issues?q=is:pr+repo:PointCloudLibrary/pcl"
+    url += "+merged:>=" + start_date
+    pr_data = list()
+    page = 1
+    total_count = 1  # for the first iteration assume that total is 1
+    while len(pr_data) < total_count:
+        response = requests.get(url + "&page=" + str(page) + "&per_page=100")
+        data = response.json()
+        if response.status_code != 200:
+            sys.exit(
+                "Failed to fetch PR info. Server response: '{}'".format(data["message"])
+            )
+        total_count = data["total_count"]
+        pr_data.extend(data["items"])
+        page += 1
+    return pr_data
+
+
+def filter_labels(labels, prefix):
+    """
+    Filter a given list of PR labels, keeping only those starting with a given prefix.
+    The prefix is stripped from the labels.
+    """
+    return [
+        label["name"][len(prefix) :]
+        for label in labels
+        if label["name"].startswith(prefix)
+    ]
+
+
+def make_pr_bullet_point(pr, prefix=None):
+    ref = "[#{0}](https://github.com/PointCloudLibrary/pcl/pull/{0})".format(
+        pr["number"]
+    )
+    if prefix == "modules" or prefix == "groups":
+        p = pr[prefix]
+    else:
+        p = []
+    prefix = "".join(["[" + k + "]" for k in p])
+    if prefix:
+        prefix = "**" + prefix + "** "
+    return "* " + prefix + pr["title"] + " [" + ref + "]"
+
+
+def generate_group_section(key, prs):
+    clog = list()
+    filtered_prs = [pr for pr in prs if key in pr["groups"]]
+    title, description = GROUPS[key]
+    if filtered_prs and title:
+        clog += [f"\n**{title}** *{description}*\n"]
+        clog += [make_pr_bullet_point(pr, "modules") for pr in filtered_prs]
+    return clog
+
+
+def generate_module_section(key, prs):
+    clog = list()
+    filtered_prs = [pr for pr in prs if key in pr["modules"]]
+    title = MODULES[key]
+    if filtered_prs and title:
+        clog += [f"\n#### {title}:\n"]
+        clog += [make_pr_bullet_point(pr, "groups") for pr in filtered_prs]
+    return clog
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="""
+    Generate changelog from pull requests merged after the latest release.
+    """,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument("--save", type=Path, help="Save PR data fetched from GitHub")
+    parser.add_argument("--load", type=Path, help="Load PR data from a file")
+    args = parser.parse_args()
+
+    if not args.load:
+        date = fetch_latest_release_date()
+        pr_data = fetch_pr_data(date)
+        if args.save:
+            with open(args.save, "w") as fp:
+                fp.write(json.dumps(pr_data))
```

Comment:
But I still want to dump for debugging purposes because GitHub limits the number of requests that a client is allowed to make.

Reply:
> dump for debugging purposes
> GitHub limits the number of requests that a client is allowed to make

This will run every merge-on-master so maybe using a cache might be a better idea. Iteratively update a cached changelog.

Reply:
GitHub allows [10 unauthenticated requests per minute](https://developer.github.com/v3/search/). It's only a problem when you are actively debugging the script and make several iterations per minute.